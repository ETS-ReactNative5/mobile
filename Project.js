import React, { Component } from 'react';
import {
  Linking,
  StyleSheet,
  Text,
  TouchableOpacity,
  View
} from 'react-native';

class Project extends Component {
  handleClick() {
    const zurl='http://zooniverse.org'
    Linking.canOpenURL(zurl).then(supported => {
      if (supported) {
        Linking.openURL(zurl);
      } else {
        console.log('Don\'t know how to open URI: ' + zurl);
      }
    });
  }

  render() {
    return (
      <View style={styles.container}>
        <TouchableOpacity
          onPress={this.handleClick}>
          <Text style={styles.project}>{this.props.title}</Text>
        </TouchableOpacity>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    borderBottomColor: 'black',
    borderBottomWidth: StyleSheet.hairlineWidth,
    marginBottom: 12,
    marginLeft: 5,
    paddingLeft: 25,
    paddingRight: 25,
    paddingBottom: 10
  },
  project: {
    color: '#F9F9F9',
    fontSize: 20,
  },
});

export default Project
