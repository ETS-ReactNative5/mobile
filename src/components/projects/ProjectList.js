import React, { Component } from 'react'
import {
  SectionList,
  View,
} from 'react-native'
import EStyleSheet from 'react-native-extended-stylesheet'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import GoogleAnalytics from 'react-native-google-analytics-bridge'
import PropTypes from 'prop-types';
import R from 'ramda'

import ProjectTile from './ProjectTile';
import NavBar from '../NavBar'
import FontedText from '../common/FontedText'
import * as navBarActions from '../../actions/navBar'
import { GLOBALS } from '../../constants/globals'
import Theme from '../../theme'

GoogleAnalytics.trackEvent('view', 'Project')

const mapStateToProps = (state, ownProps) => {
    const { selectedProjectTag } = ownProps;
    const inPreviewMode = selectedProjectTag === 'preview'
    let projectList
    
    // Grab all of the projects from the selected Project Tag
    if (selectedProjectTag === 'recent') {
      const activeProjects = R.filter((project) => { return project.activity_count > 0 }, state.user.projects);
      projectList = state.projects.projectList.filter((project) => R.keys(activeProjects).includes(project.id));
    } else if (inPreviewMode) {
        projectList = state.projects.previewProjectList
    }    
    else {
        projectList = state.projects.projectList.filter((project) => R.contains(selectedProjectTag, project.tags))
    }

    // Seperate out the native workflows and non-native workflows    
    const swipeEnabledProjects = projectList.filter((project) => R.any((workflow) => workflow.swipe_verified)(project.workflows))
    const nonSwipeEnabledProjects = projectList.filter((project) => R.all((workflow) => !workflow.swipe_verified)(project.workflows))

    return {
        swipeEnabledProjects,
        nonSwipeEnabledProjects,
        promptForWorkflow: state.main.settings.promptForWorkflow || false,
        isLoading: state.projects.isLoading,
        collaboratorIds: state.projects.collaboratorIds,
        ownerIds: state.projects.ownerIds,
        inPreviewMode
    };
}

const mapDispatchToProps = (dispatch) => ({
    navBarActions: bindActionCreators(navBarActions, dispatch)
})

const PAGE_KEY = 'ProjectList';

class ProjectList extends Component {
    
    static renderNavigationBar() {
        return <NavBar pageKey={PAGE_KEY} showBack={true} />;
    }

    componentDidMount() {
        const title = GLOBALS.DISCIPLINES.find((element) => element.value === this.props.selectedProjectTag).label
        this.props.navBarActions.setTitleForPage(title, PAGE_KEY);

        if (this.props.inPreviewMode) {
            this.props.navBarActions.setNavbarColorForPage(Theme.$testRed, PAGE_KEY)
        } else {
            this.props.navBarActions.setNavbarColorForPageToDefault(PAGE_KEY)
        }
    }

    _emptyText() {
        if (!this.props.isLoading) {
            return 'Sorry, but you have no mobile friendly projects to display'
        } else {
            return 'Loading Projects...'
        }
    }
    
    _seperatorHeightStyle(data) {
        return {height: !!data.leadingItem && !!data.trailingSection ? 50 : 0}
    }

    render() {
        const {inPreviewMode, ownerIds, collaboratorIds, swipeEnabledProjects, nonSwipeEnabledProjects } = this.props
        let sections = []
        if (inPreviewMode) {
            if (!R.isEmpty(ownerIds)) {
                const ownerProjects = swipeEnabledProjects.filter((project) => ownerIds.includes(project.id))
                sections.push({data: ownerProjects, title: 'Your Projects'})
            }

            if (!R.isEmpty(collaboratorIds)) {
                const collaboratorProjects = swipeEnabledProjects.filter((project) => collaboratorIds.includes(project.id))
                sections.push({data: collaboratorProjects, title: 'Collaborations'})
            }
        } else {
            if (!R.isEmpty(swipeEnabledProjects)) {
                sections.push({data: swipeEnabledProjects, title: 'Made For Mobile'});
            }
    
            if (!R.isEmpty(this.props.nonSwipeEnabledProjects)) {
                sections.push({data: nonSwipeEnabledProjects, title: 'In-Browser Experience'})
            }
        }
        

        return (
            <SectionList
                contentContainerStyle={styles.contentContainer}
                stickySectionHeadersEnabled={false}
                ItemSeparatorComponent={() => <View style={styles.separatorView} />}
                SectionSeparatorComponent={(data) => <View style={this._seperatorHeightStyle(data)} />}
                renderItem={({item}) => <ProjectTile project={item} inPreviewMode={this.props.inPreviewMode}/>}
                renderSectionHeader={({section}) => <FontedText style={styles.sectionHeader}> { section.title } </FontedText>}
                sections={sections}
                ListEmptyComponent={() => <FontedText style={styles.emptyComponent}> {this._emptyText()} </FontedText>}
                keyExtractor={(item, index) => index}
            />
        );
    }
}

const styles = EStyleSheet.create({
    contentContainer: {
        paddingBottom: 25,
        paddingTop: 35
    },
    separatorView: {
        height: 25
    },
    sectionHeader: {
        fontSize: 26, 
        marginLeft: 25, 
        marginBottom: 20, 
        fontWeight: 'bold',
        color: '$headerGrey'
    },
    emptyComponent: {
        fontStyle: 'italic', 
        marginHorizontal: 20, 
        color: '$headerGrey'
    }

});

ProjectList.propTypes = {
    swipeEnabledProjects: PropTypes.array,
    nonSwipeEnabledProjects: PropTypes.array,
    recentsList: PropTypes.array,
    promptForWorkflow: PropTypes.bool,
    isLoading: PropTypes.bool,
    isSuccess: PropTypes.bool,
    selectedProjectTag: PropTypes.string,
    navBarActions: PropTypes.any,
    inPreviewMode: PropTypes.bool,
    collaboratorIds: PropTypes.array,
    ownerIds: PropTypes.array
}

export default connect(mapStateToProps, mapDispatchToProps)(ProjectList)