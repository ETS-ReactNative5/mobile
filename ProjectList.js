import React from 'react'
import {
  Dimensions,
  ListView,
  StyleSheet,
  Text,
  View
} from 'react-native'
import EStyleSheet from 'react-native-extended-stylesheet'
import Orientation from 'react-native-orientation'
import apiClient from 'panoptes-client/lib/api-client'
import Project from './Project'
import {MOBILE_PROJECTS} from './constants/mobile_projects'

var {height, width} = Dimensions.get('window');

var ProjectList = React.createClass({
  getInitialState: function() {
    return {
      dataSource: new ListView.DataSource({
        rowHasChanged: (row1, row2) => row1 !== row2,
      })
    };
  },

  componentDidMount() {
    this.getProjects()
    Orientation.lockToPortrait
  },

  getDataSource(projects: Array<any>): ListView.DataSource {
    return this.state.dataSource.cloneWithRows(projects);
  },

  getProjects(){
    apiClient.type('projects').get({id: MOBILE_PROJECTS, cards: true, sort: 'display_name'})
      .then((projects) => {
        this.setState({
          dataSource: this.getDataSource(projects),
        });
      })
      .catch((error) => {
        console.error(error);
      })
  },

  renderRow(project) {
    return (
      <Project project={project}/>
    );
  },

  render() {
    return (
      <View style={styles.container}>
        <View style={styles.titleContainer}>
          <Text style={styles.title}>
            Zooniverse
          </Text>
        </View>
        <View style={styles.subtitleContainer}>
          <Text style={styles.subtitle}>
            Mobile-friendly Projects
          </Text>
        </View>
        <ListView
          dataSource={this.state.dataSource}
          renderRow={this.renderRow}
        />
      </View>
    );
  }
});

const styles = EStyleSheet.create({
  container: {
    flex: 1
  },
  titleContainer: {
    backgroundColor: '$backgroundColor',
    justifyContent: 'center',
    paddingBottom: 10,
    paddingTop: 32,
    alignItems: 'center',
    width: width
  },
  title: {
    color: '$textColor',
    fontSize: 22
  },
  subtitleContainer: {
    borderBottomColor: '$borderColor',
    borderBottomWidth: StyleSheet.hairlineWidth,
    marginBottom: 15
  },
  subtitle: {
    color: '$textColor',
    fontSize: 24,
    padding: 15,
  }
});

module.exports = ProjectList
