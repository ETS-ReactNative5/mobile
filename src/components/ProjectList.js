import React from 'react'
import {
  Alert,
  ListView,
  View
} from 'react-native'
import EStyleSheet from 'react-native-extended-stylesheet'
import StyledText from './StyledText'
import { MOBILE_PROJECTS } from '../constants/mobile_projects'
import Project from './Project'
import NavBar from './NavBar'
import apiClient from 'panoptes-client/lib/api-client'
import { connect } from 'react-redux'
import GoogleAnalytics from 'react-native-google-analytics-bridge'

GoogleAnalytics.trackEvent('view', 'Project')

const mapStateToProps = (state) => ({
  user: state.user,
  isConnected: state.isConnected
})

class ProjectList extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      dataSource: new ListView.DataSource({
        rowHasChanged: (row1, row2) => row1 !== row2,
      })
    }
  }

  componentDidMount() {
    if (this.state.dataSource.getRowCount() === 0){
      this.getProjects()
    }
  }

  getDataSource(projects: Array<any>): ListView.DataSource {
    return this.state.dataSource.cloneWithRows(projects);
  }

  getProjects(){
    apiClient.type('projects').get({id: MOBILE_PROJECTS, cards: true, tags: this.props.tag, sort: 'display_name'})
      .then((projects) => {
        this.setState({
          dataSource: this.getDataSource(projects),
        });
      })
      .catch((error) => {
        Alert.alert(
          'Error',
          'The following error occurred.  Please close down Zooniverse and try again.  If it persists please notify us.  \n\n' + error,
        )
      })
  }

  renderRow(project) {
    return (
      <Project project={project}/>
    );
  }

  static renderNavigationBar() {
    return <NavBar title={"Projects"} showBack={true} />;
  }

  render() {
    const projectList =
      <ListView
        dataSource={this.state.dataSource}
        renderRow={this.renderRow}
        enableEmptySections={true}
      />

    const noConnection =
      <View style={styles.messageContainer}>
        <StyledText textStyle={'errorMessage'}
          text={'You must have an internet connection to use Zooniverse Mobile'} />
      </View>

    return (
      <View style={styles.container}>
        <View style={styles.innerContainer}>
          { this.props.isConnected ? projectList : noConnection }
        </View>
      </View>
    );
  }
}

const styles = EStyleSheet.create({
  container: {
    flex: 1,
    paddingTop: 60,
  },
  innerContainer: {
    flex: 1,
    marginTop: 30
  },
  listStyle: {
    paddingTop: 90
  },
  messageContainer: {
    padding: 15,
  },
});

ProjectList.propTypes = {
  user: React.PropTypes.object,
  isConnected: React.PropTypes.bool,
  tag: React.PropTypes.string
}

export default connect(mapStateToProps)(ProjectList)
