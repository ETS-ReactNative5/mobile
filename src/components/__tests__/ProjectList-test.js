import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { ProjectList } from '../ProjectList'
import Project from '../Project'

const project = {
  avatar_src: 'fake_avatar',
  display_name: 'Nice project'
}

jest.mock('ListView', () => require('react').createClass({
    statics: {
        DataSource: require.requireActual('ListView').DataSource,
    },
    render() {
        return <Project project={project} />
    },
}))

it('renders correctly', () => {
  const tree = renderer.create(
    <ProjectList isConnected={true} fetchProjects={jest.fn} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
