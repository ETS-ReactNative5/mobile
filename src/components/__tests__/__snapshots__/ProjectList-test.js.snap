exports[`test renders correctly 1`] = `
<View
  style={undefined}>
  <View
    style={undefined}>
    <View
      style={undefined}>
      <Image
        source={
          Object {
            "uri": "https://fake_avatar",
          }
        }
        style={undefined} />
      <TouchableOpacity
        activeOpacity={0.5}
        onPress={[Function]}
        style={undefined}>
        <View
          style={undefined}>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            numberOfLines={1}
            style={undefined}>
            Nice project
          </Text>
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontFamily": "FontAwesome",
                  "fontSize": 12,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                undefined,
              ]
            }>
            ï„…
          </Text>
        </View>
      </TouchableOpacity>
    </View>
  </View>
</View>
`;
